### PR TITLE
Implement tabular modification

### PR DESCRIPTION
**Pull request by etiennehomer**

### CODE REVIEW

File: src/main/java/org/gridsuite/modification/server/dto/TabularModificationInfos.java [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+import java.util.List;
+
+/**
+ * @author Etienne Homer <etienne.homer at rte-france.com>
+ */
+@SuperBuilder
+@NoArgsConstructor
+@Getter
+@Setter
+@ToString(callSuper = true)
```
Comment (Tristan-WorkGH) [inline src/main/java/org/gridsuite/modification/server/dto/TabularModificationInfos.java:35]:
```suggestion
@Data(callSuper = true)
```
You can replace by `lombok.Data` to be short

**Review by Tristan-WorkGH — CHANGES_REQUESTED** [review]
Will look more in details later

File: src/main/java/org/gridsuite/modification/server/dto/TabularModificationInfos.java (hunk 1 above)
Comment (etiennehomer) [reply src/main/java/org/gridsuite/modification/server/dto/TabularModificationInfos.java]: done

File: src/main/java/org/gridsuite/modification/server/modifications/TabularModification.java [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+    @Override
+    public void check(Network network) throws NetworkModificationException {
+        if (modificationInfos == null) {
+            throw new NetworkModificationException(TABULAR_MODIFICATION_ERROR, "No tabular modification to apply !!");
+        }
+    }
+
+    @Override
+    public void apply(Network network, Reporter subReporter) {
+        modificationInfos.getModifications().forEach(modification -> modification.toModification().apply(network));
```
Comment (FranckLecuyer) [inline src/main/java/org/gridsuite/modification/server/modifications/TabularModification.java:37]: Add the subReporter as second argument to the apply method call ?

File: src/main/java/org/gridsuite/modification/server/service/NetworkModificationService.java [hunk 3]
```diff
@@ -212,7 +212,7 @@ public Optional<NetworkModificationResult> moveModifications(UUID groupUuid, UUI
         List<ModificationInfos> movedModifications = networkModificationRepository.moveModifications(groupUuid, originGroupUuid, modificationsToMove, before)
             .stream()
             .filter(m -> !m.getStashed())
-            .map(ModificationEntity::toModificationInfos)
+            .map(networkModificationRepository::getModificationInfos)
```
Comment (FranckLecuyer) [inline src/main/java/org/gridsuite/modification/server/service/NetworkModificationService.java:215]: Change to be made also in saveAndApplyModifications method ?

**Review by FranckLecuyer — CHANGES_REQUESTED** [review]

File: src/main/java/org/gridsuite/modification/server/service/NetworkModificationService.java (hunk 3 above)
Comment (etiennehomer) [reply src/main/java/org/gridsuite/modification/server/service/NetworkModificationService.java:215]: I didn't do it cause it was raising an error on the duplication of a tabular modification gathering generator modifications.

I still have the N+1 problem here (it's the only piece of code where I have it I think). Ok I'll fix it

Comment (etiennehomer) [reply src/main/java/org/gridsuite/modification/server/service/NetworkModificationService.java:215]: It is fixed

**Review by FranckLecuyer — APPROVED** [review]

**Review by Tristan-WorkGH — APPROVED** [review]

File: src/main/java/org/gridsuite/modification/server/modifications/TabularModification.java [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+        if (modificationInfos == null) {
+            throw new NetworkModificationException(TABULAR_MODIFICATION_ERROR, "No tabular modification to apply !!");
+        }
+    }
+
+    @Override
+    public void apply(Network network, Reporter subReporter) {
+        AtomicReference<Integer> applicationFailuresCount = new AtomicReference<>(0);
+        modificationInfos.getModifications().forEach(modification -> {
+            try {
```
Comment (etiennehomer) [inline src/main/java/org/gridsuite/modification/server/modifications/TabularModification.java:46]: catch application fail exception

File: src/main/java/org/gridsuite/modification/server/modifications/TabularModification.java [hunk 5]
```diff
@@ -34,24 +41,37 @@ public void check(Network network) throws NetworkModificationException {
...
     @Override
     public void apply(Network network, Reporter subReporter) {
-        modificationInfos.getModifications().forEach(modification -> modification.toModification().apply(network));
-        String defaultMessage;
+        AtomicReference<Integer> applicationFailuresCount = new AtomicReference<>(0);
+        modificationInfos.getModifications().forEach(modification -> {
+            try {
+                modification.toModification().apply(network);
+            } catch (NetworkModificationException e) {
+                applicationFailuresCount.set(applicationFailuresCount.get() + 1);
```
Comment (FranckLecuyer) [inline src/main/java/org/gridsuite/modification/server/modifications/TabularModification.java:49]: applicationFailuresCount.incrementAndGet()

File: src/test/java/org/gridsuite/modification/server/modifications/tabularmodifications/TabularGeneratorModificationsTest.java [hunk 6]
```diff
@@ -40,7 +40,8 @@ protected ModificationInfos buildModification() {
         List<ModificationInfos> modifications = List.of(
                 GeneratorModificationInfos.builder().equipmentId("idGenerator").maxActivePower(new AttributeModification<>(500., OperationType.SET)).build(),
                 GeneratorModificationInfos.builder().equipmentId("v5generator").maxActivePower(new AttributeModification<>(500., OperationType.SET)).build(),
-                GeneratorModificationInfos.builder().equipmentId("v6generator").maxActivePower(new AttributeModification<>(500., OperationType.SET)).build()
+                GeneratorModificationInfos.builder().equipmentId("v6generator").maxActivePower(new AttributeModification<>(500., OperationType.SET)).build(),
+                GeneratorModificationInfos.builder().equipmentId("unknownGenerator").maxActivePower(new AttributeModification<>(500., OperationType.SET)).build()
         );
         return TabularModificationInfos.builder()
                 .modificationType("GENERATOR_MODIFICATION")
```
Comment (FranckLecuyer) [inline src/test/java/org/gridsuite/modification/server/modifications/tabularmodifications/TabularGeneratorModificationsTest.java:47]: Add an assertLogMessage in assertAfterNetworkModificationCreation method, in order to check that the report for the not found 'unknownGenerator' and that the report for the modifications count have been generated

File: src/main/java/org/gridsuite/modification/server/modifications/TabularModification.java [hunk 7]
```diff
@@ -34,24 +41,37 @@ public void check(Network network) throws NetworkModificationException {
 
     @Override
     public void apply(Network network, Reporter subReporter) {
-        modificationInfos.getModifications().forEach(modification -> modification.toModification().apply(network));
-        String defaultMessage;
+        AtomicReference<Integer> applicationFailuresCount = new AtomicReference<>(0);
```
[the remark this answers is not in the document]
Comment (etiennehomer) [reply src/main/java/org/gridsuite/modification/server/modifications/TabularModification.java]: done

File: src/main/java/org/gridsuite/modification/server/modifications/TabularModification.java (hunk 5 above)
Comment (etiennehomer) [reply src/main/java/org/gridsuite/modification/server/modifications/TabularModification.java]: done

File: src/test/java/org/gridsuite/modification/server/modifications/tabularmodifications/TabularGeneratorModificationsTest.java (hunk 6 above)
Comment (etiennehomer) [reply src/test/java/org/gridsuite/modification/server/modifications/tabularmodifications/TabularGeneratorModificationsTest.java:49]: done

File: src/main/java/org/gridsuite/modification/server/modifications/TabularModification.java [hunk 8]
```diff
@@ -34,24 +41,37 @@ public void check(Network network) throws NetworkModificationException {
...
+        modificationInfos.getModifications().forEach(modification -> {
+            try {
+                modification.toModification().apply(network);
+            } catch (NetworkModificationException e) {
+                applicationFailuresCount.set(applicationFailuresCount.get() + 1);
+                subReporter.report(Report.builder()
+                        .withKey("generatorModification" + applicationFailuresCount.get())
```
[the remark this answers is not in the document]
Comment (etiennehomer) [reply src/main/java/org/gridsuite/modification/server/modifications/TabularModification.java]: done

**Review by FranckLecuyer — APPROVED** [review]

**Pull request merged**